dotnet/TagExpressions/TagToken.cs

```diff
@@ -0,0 +1,22 @@
+internal enum TagTokenType
+{
+    Identifier,
+    And,
+    Or,
+    Not,
+    LParen,
+    RParen,
+    End
+}
+
+internal class TagToken
+{
+    public TagTokenType Type { get; }
```

Comment:
From other similar project experience, I have found it very useful to store the position of the token value as well withing the original expression. (I.e. the character index where this particular token has been found). This can be used to provide a better error message in case the error is related to a token. Like:

```
Invalid operation "and": "@foo and and @bar"
                                   ^
```